### ISSUE TITLE
Unable to scrobble (users without PlexPass)

### ISSUE THREAD

**Issue by fuzeman** [trakt/Plex-Trakt-Scrobbler#160]
In the 0.9.* update the method of retrieving watching activity was changed, this means some operating systems need support specifically added.

Currently, the following operating systems are known to be supported:
 - Darwin / OS X
 - Linux
    - Ubuntu 12.04
 - Windows
   - XP (?)
   - 7
   - 8

-----------------------------------------
**Operating system not supported?**

1 - In a terminal/shell, enter the following command:
*(make sure you have python installed)*

```bash
python -c "import platform; print (platform.system(), platform.linux_distribution())"
```

2 - Locate the path to ```Plex Media Server.log```, [this page](https://support.plex.tv/hc/en-us/articles/200250417-Plex-Media-Server-Log-Files) will give you a hint *(double-check the path listed is correct)*

3 - Post a comment below with:

 - Output from the command at step 1
 - Path to ```Plex Media Server.log``` on your system

**Comment by shotfirer** [comment]
QNAP TS-219p (ARM)
I have two Python installation.

Python 2.5.6:
`# python -c "import platform; print (platform.system(), platform.linux_distribution())"   
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: 'module' object has no attribute 'linux_distribution'`
Python 2.7.3:
`# python2 -c "import platform; print (platform.system(), platform.linux_distribution())"  
('Linux', ('', '', ''))`

Path to `Plex Media Server.log`:
`/share/HDA_DATA/.qpkg/PlexMediaServer/Library/Plex\ Media\ Server/Logs/Plex\ Media\ Server.log`

**Comment by BrutuZ** [comment]
Does that mean "Logging (Legacy)" no longer works?
I have a PlexPass but have been using the Logging method because WebSocket doesn't get some info from the LG Client (which sends HTTP requests directly to the server), specifically IP Address returns None, so I ca no longer use the IP filter.

**Plex Media Server.log**
```Jan 10, 2015 17:31:12:659 [9956] DEBUG - Request: [192.168.0.13:59990] GET /:/timeline?time=19676&duration=2553846&state=playing&key=%2Flibrary%2Fmetadata%2F14072&ratingKey=14072&containerKey=%2Flibrary%2Fmetadata%2F14072&guid=com.plexapp.agents.thetvdb%3A%2F%2F255326%2F2%2F7%3Flang%3Den (14 live)```

**com.plexapps.plugin.trakttv.log**
```2015-01-10 17:31:12,688 (275c) :  DEBUG (logkit:13) - [pyemitter] [plex_activity.sources.s_websocket.main.WebSocket]:emit(event: 'websocket.playing', args: ({u'viewOffset': 19676, u'url': u'', u'ratingKey': u'14072', u'_elementType': u'PlaySessionStateNotification', u'state': u'playing', u'sessionKey': u'1', u'key': u'/library/metadata/14072', u'guid': u'com.plexapp.agents.thetvdb://255326/2/7?lang=en'},), kwargs: {})
2015-01-10 17:31:12,690 (275c) :  DEBUG (logkit:13) - [pyemitter] [plex_activity.activity.Activity   ]:emit(event: 'websocket.playing', args: ({u'viewOffset': 19676, u'url': u'', u'ratingKey': u'14072', u'_elementType': u'PlaySessionStateNotification', u'state': u'playing', u'sessionKey': u'1', u'key': u'/library/metadata/14072', u'guid': u'com.plexapp.agents.thetvdb://255326/2/7?lang=en'},), kwargs: {})
2015-01-10 17:31:12,693 (275c) :  INFO (logkit:16) - [plugin.pts.scrobbler] Ignoring item [14072](Defiance) played by filtered "filter_networks": None```

**Comment by fuzeman** [comment]
@shotfirer the develop branch includes support for that system now (download: [develop.zip](https://github.com/trakt/Plex-Trakt-Scrobbler/archive/develop.zip)), it will be included in a new beta release later today.

--------------------

@BrutuZ "Logging (Legacy)" should work if your operating system is supported, just need to provide the details listed above to get it supported.

I've just moved the module that handles the activity system to a separate library ([plex.activity.py](https://github.com/fuzeman/plex.activity.py)), doing this meant it couldn't easily locate the path of "Plex Media Server.log" without ["path hints"](https://github.com/fuzeman/plex.activity.py/blob/develop/plex_activity/sources/s_logging/main.py#L16) though.

This change means applications/plugins can easily use the same activity library. *(I might build a good last.fm scrobbler plugin soon..)*

**Comment by BrutuZ** [comment]
@fuzeman  Got a look at the path hints and found out why it doesn't work even though I'm using Windows 7 64-bit. I don't use the default path at Local AppData, my Plex Data folder is in the same folder Plex itself is installed on another hard-drive.

Isn't there a way to get this value from the server settings? If not, maybe put a field within the settings dialog so that users can copy\paste it from the server settings page to avoid changing the code to put one's own custom path?

**Comment by fuzeman** [comment]
Hmm, I'll come up with another solution.

Would rather not add another option to the settings dialog unless it's absolutely necessary.

**Comment by BrutuZ** [comment]
Got it to parse the log by changing the function to ```lambda: os.path.realpath(os.getcwd() + '/../../../Logs/')``` but something is not right, the plugin is filtering the playback even though it is from a whitelisted IP. http://pastebin.com/qj0xejJR

**Comment by shotfirer** [comment]
@fuzeman Got my scrobbling back. Cool!

**Comment by fuzeman** [comment]
**v0.9.1.7-beta** is available: https://forums.plex.tv/index.php/topic/102818-rel-trakt/page-28#entry820611

@BrutuZ the "Plex Media Server.log" path is now discovered the same as past releases. I've also fixed an issue filtering Plex/Web clients (that was causing the issue you mentioned).

**Comment by BrutuZ** [comment]
Working like a charm now @fuzeman

**Comment by shotfirer** [comment]
No scrobbling again. It was working in 9.1.6-dev, but with 0.9.1.7 it is not. And again, there's nothing in com.plexapps.plugin.trakttv.log when i'm watching something.

**Comment by fuzeman** [comment]
Does this still happen after restarting PMS (so the plugin restarts)?

It's a bit hard to work out what's happened from your log files, I can't really see why the scanner stopped processing your "Plex Media Server.log" file.

---------------------

I've noticed there can also be some lag in the processing of logs on some systems, so I'll look at reducing the processing delay (or add an option to the preferences) - doing this can increase IO/CPU usage though.

**Comment by shotfirer** [comment]
Stupid me 8) I'd always just replaced plugin, without restarting PMS. Previous versions just appeared in the system and all features started to work, but not this time, i think.
Restart fixed it, at least for now. I'll check it for a day or two, than report here.
Thank you.

**Comment by shotfirer** [comment]
And it's not working again. This time it seems, that scrobbling was broken by log rotation.
```
2015-01-17 22:41:15,689 (-4c1e9b70) :  DEBUG (logging_handler:39) - [plugin.pts.scrobbler] total_progress: 0.307200189142 (view_offset: 790000, duration: 2571613)
2015-01-17 22:41:15,745 (-4c1e9b70) :  DEBUG (logging_handler:39) - [plugin.pts.scrobbler_logging] [PL 31] Nothing to do this time for Stargate Atlantis ('tvdb', '70851') [u: None, c: None, s: 'TV Shows', a: '192.168.0.115']
2015-01-17 22:41:46,345 (-4c1e9b70) :  DEBUG (logging_handler:39) - [plex_activity.sources.s_logging.main] Log file moved (probably rotated), closing
```
And that's it, no scrobbling past this point.
Complete logs here:
https://www.dropbox.com/sh/4vuux8fsvvn944q/AADNnEX8SpjajIcvzdI8L6tXa?dl=0

**Comment by shotfirer** [comment]
Yeah, that's definitely the issue here. After PMS restart the scrobbling works fine, but only until first log rotation.

**Comment by fuzeman** [comment]
@shotfirer I've discovered errors aren't logged correctly from the activity modules, could you grab **v0.9.1.9-beta** from https://forums.plex.tv/index.php/topic/102818-rel-trakt/page-31#entry829615 and provide an updated log file.

**Comment by shotfirer** [comment]
I suppose this is what you've looking for:
```
2015-01-21 22:54:39,069 (-4c192b70) :  DEBUG (logging_handler:46) - [plex_activity.sources.s_logging.main] Log file moved (probably rotated), closing
2015-01-21 22:54:39,205 (-4c192b70) :  ERROR (logging_handler:46) - [plex_activity.sources.base] Exception raised in "logging" activity source
Traceback (most recent call last):
  File "/root/Library/Plex Media Server/Plug-ins/Trakttv.bundle/Contents/Libraries/Shared/plex_activity/sources/base.py", line 22, in _run_wrapper
    self.run()
  File "/root/Library/Plex Media Server/Plug-ins/Trakttv.bundle/Contents/Libraries/Shared/plex_activity/sources/s_logging/main.py", line 73, in run
    line = self.read_line_retry(ping=True)
  File "/root/Library/Plex Media Server/Plug-ins/Trakttv.bundle/Contents/Libraries/Shared/plex_activity/sources/s_logging/main.py", line 121, in read_line_retry
    self.close()
  File "/root/Library/Plex Media Server/Plug-ins/Trakttv.bundle/Contents/Libraries/Shared/plex_activity/sources/s_logging/main.py", line 138, in close
    self.file.close()
  File "/root/Library/Plex Media Server/Plug-ins/Trakttv.bundle/Contents/Libraries/Shared/asio/file.py", line 89, in close
    return self.get_handler().close(self)
  File "/root/Library/Plex Media Server/Plug-ins/Trakttv.bundle/Contents/Libraries/Shared/asio/interfaces/posix.py", line 105, in close
    os.close(fp.fd)
OSError: [Errno 9] Bad file descriptor
2015-01-21 22:54:39,578 (-4c192b70) :  DEBUG (logging_handler:46) - [raven.base.Client] Sending message of length 1988 to requests+http://sentry.skipthe.net/api/1/store/
2015-01-21 22:54:39,607 (-4c192b70) :  INFO (logging_handler:46) - [requests.packages.urllib3.connectionpool] Starting new HTTP connection (1): sentry.skipthe.net

```

**Comment by fuzeman** [comment]
@shotfirer could you test the latest development release: https://github.com/trakt/Plex-Trakt-Scrobbler/archive/develop.zip

**Comment by shotfirer** [comment]
@fuzeman So far so good. It seems, that whatever you did fixed it. Still some warnings in the log, but no errors and scrobbling was not broken by log rotation. Will check it for a few days though. Great job, thank you, guys!
**com.plexapps.plugin.trakttv.log**
```
2015-01-22 23:58:22,446 (-4c19bb70) :  DEBUG (logging_handler:46) - [plex_activity.sources.s_logging.main] Log file moved (probably rotated), closing
2015-01-22 23:58:22,479 (-4c19bb70) :  WARNING (logging_handler:46) - [plex_activity.sources.s_logging.main] file.close() - ignoring raised exception: [Errno 9] Bad file descriptor (already closed)
2015-01-22 23:58:23,496 (-4c19bb70) :  INFO (logging_handler:46) - [plex_activity.sources.s_logging.main] Opened file path: "/share/HDA_DATA/.qpkg/PlexMediaServer/Library/Plex Media Server/Logs/Plex Media Server.log"
2015-01-22 23:58:25,586 (-4c19bb70) :  DEBUG (logging_handler:46) - [plex_activity.sources.s_logging.parsers.base] break on "Jan 22, 2015 23:58:25 [0xb148f430] DEBUG - Request: [192.168.0.115:50885] GET /:/progress?key=3657&identifier=com.plexapp.plugins.library&time=970000&state=playing (3 live)"
2015-01-22 23:58:25,595 (-4c19bb70) :  WARNING (logging_handler:46) - [plex_activity.sources.s_logging.parsers.now_playing] Invalid activity match, missing key state ({'state': None, 'identifier': 'com.plexapp.plugins.library', 'ratingKey': '3657', 'key': '3657', 'time': None})
2015-01-22 23:58:31,496 (-4c19bb70) :  DEBUG (logging_handler:46) - [plex_activity.sources.s_logging.parsers.base] break on "Jan 22, 2015 23:58:30 [0xb3a16430] VERBOSE - Finished writing response for GET /:/progress?key=3657&identifier=com.plexapp.plugins.library&time=975000&state=playing, 0 bytes in 325ms"
```
I think, that the cause of the second warning, is that the log was rotated in the middle of event.
Here's the beginning of the new log file.
**Plex Media Server.log**
```
Jan 22, 2015 23:58:19 [0xb148f430] DEBUG -  * time => 964000
Jan 22, 2015 23:58:19 [0xb148f430] DEBUG -  * state => playing
Jan 22, 2015 23:58:19 [0xb148f430] DEBUG - Play progress on 3657 'Snakes and Whatnot' - got played 964000 ms by account 1!
Jan 22, 2015 23:58:19 [0xb3a16430] VERBOSE - Finished writing response for GET /:/progress?key=3657&identifier=com.plexapp.plugins.library&time=964000&state=playing, 0 bytes in 676ms
Jan 22, 2015 23:58:19 [0xb148f430] VERBOSE - Completed request: GET /:/progress?key=3657&identifier=com.plexapp.plugins.library&time=964000&state=playing: 2200
```

**Comment by fuzeman** [comment]
@shotfirer that fix has now been released in **v0.9.1.10**: https://forums.plex.tv/index.php/topic/102818-rel-trakt/page-32#entry831576
